src/internal/pjs/server_internal_test.go

```diff
@@ -0,0 +1,480 @@
+	})
+}
+
+func TestWalkJob(t *testing.T) {
+	c, fc := setupTest(t)
+	ctx := pctx.TestContext(t)
+	depth := 3
+	fullBinaryJobTree(t, ctx, depth, c, fc)
+
+	expectedSize := int64(math.Pow(2, float64(depth)) - 1)
+	expected := make(map[int64][]int64)
+	for i := int64(1); i <= expectedSize; i++ {
+		left := 2 * i
+		if left <= expectedSize {
+			expected[i] = append(expected[i], left)
+		}
+		right := 2*i + 1
+		if right <= expectedSize {
+			expected[i] = append(expected[i], right)
+
+		}
+	}
+	walkC, err := c.WalkJob(ctx, &pjs.WalkJobRequest{
+		Job: &pjs.Job{
+			Id: 1,
+		},
+		Algorithm: pjs.WalkAlgorithm_LEVEL_ORDER,
+	})
+	require.NoError(t, err)
+	actual := make(map[int64][]int64)
+	for {
+		resp, err := walkC.Recv()
+		if err != nil {
+			if errors.As(err, io.EOF) {
+				break
+			}
+			require.NoError(t, err)
+		}
+		parentJob := resp.Details.JobInfo.ParentJob
+		if parentJob != nil && parentJob.Id != 0 {
+			actual[parentJob.Id] = append(actual[parentJob.Id], resp.Id.Id)
+		}
+		require.NoError(t, err)
+	}
+	require.NoDiff(t, expected, actual, nil)
+}
+
+// runJob does work through PJS.
+func runJob(t *testing.T, ctx context.Context, c pjs.APIClient, in *pjs.CreateJobRequest,
+	fn func(resp *pjs.ProcessQueueResponse) error) (*pjs.JobInfo_Success, error) {
+	jres, err := c.CreateJob(ctx, in)
+	require.NoError(t, err)
+	return runJobFrom(t, ctx, c, jres.Id.Id, in.Program, fn)
+}
+
+func runJobFrom(t *testing.T, ctx context.Context, c pjs.APIClient, from int64, programStr string,
+	fn func(resp *pjs.ProcessQueueResponse) error) (*pjs.JobInfo_Success, error) {
+	program, err := fileset.ParseID(programStr)
+	require.NoError(t, err)
+	programHash := []byte(program.HexString())
+	ctx, cf := context.WithCancel(ctx)
+	defer cf()
+	eg, ctx := errgroup.WithContext(ctx)
+	eg.Go(func() error {
+		ctx, cf := context.WithCancel(ctx)
+		defer cf()
+		pqc, err := c.ProcessQueue(ctx)
+		require.NoError(t, err)
+		err = processQueue(pqc, programHash, fn)
+		if errors.Is(err, io.EOF) {
+			return nil
+		}
+		if status.Code(err) == codes.Canceled {
+			err = nil
+		}
+		return err
+	})
+	var ret *pjs.JobInfo_Success
+	eg.Go(func() error {
+		jobInfo, err := await(ctx, c, from)
+		if err != nil {
+			return err
+		}
+		ret = jobInfo.GetSuccess()
+		cf() // success, cancel the other gorountine
+		return nil
+	})
+	err = eg.Wait()
+	require.NoError(t, err)
+	return ret, nil
+}
+
+func processQueue(pqc pjs.API_ProcessQueueClient, programHash []byte, fn func(resp *pjs.ProcessQueueResponse) error) error {
+	if err := pqc.Send(&pjs.ProcessQueueRequest{
+		Queue: &pjs.Queue{Id: programHash},
+	}); err != nil {
+		return err
+	}
+	for {
+		msg, err := pqc.Recv()
+		if err != nil {
+			return err
+		}
+		err = fn(msg)
+		if err != nil {
+			if sendErr := pqc.Send(&pjs.ProcessQueueRequest{
+				Result: &pjs.ProcessQueueRequest_Failed{
+					Failed: true,
+				},
+			}); sendErr != nil {
+				return sendErr
+			}
+		} else {
```

Comment:
A return on the line above is preferred to indenting the non-error case: https://go.dev/wiki/CodeReviewComments#indent-error-flow